### PR TITLE
Adds LlamaForQuestionAnswering class in modeling_llama.py along with AutoModel Support 

### PR DESCRIPTION
**Pull request by nakranivaibhav**
# What does this PR do?
Adds AutoModelForQuestionAnswering support for llama.

Fixes # (issue)
#28265

## Who can review?
@ArthurZucker @NielsRogge 

I haven't added a copy statement. When I try to add the following copy statement
`# Copied from transformers.models.bloom.modeling_bloom.BloomForQuestionAnswering.__init__ with Bloom->LLama`
It throws an error even though both the init functions are same.

When I try to make fix-copies. The fix copy wants to capitalize both the L's in Llamamodel on this line
`self.transformer = LlamaModel(config)` -> `self.transformer = LLamaModel(config`

Perhaps I am missing something and @ArthurZucker can guide me further.

I tried training the model on a subset of SQUAD and it does train fine.

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks for the PR.

This should work:
```diff
-# Copied from transformers.models.bloom.modeling_bloom.BloomForQuestionAnswering.__init__ with Bloom->LLama
+# Copied from transformers.models.bloom.modeling_bloom.BloomForQuestionAnswering.__init__ with Bloom->Llama

**Comment by nakranivaibhav** [comment]
@ArthurZucker Alright the capitalization in the copy statement affects the test. Thank you for pointing it out. @ArthurZucker

**Comment by ArthurZucker** [comment]
Now let’s make sure the CIs are green! Feel free to rebase on main!

**Comment by nakranivaibhav** [comment]
@ArthurZucker I did rebase on main, still the tests are failing. What's wrong?

P.S: I make stupid mistakes. Do Forgive

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -1413,3 +1418,100 @@ def forward(
...
+
+@add_start_docstrings(
+    """
+The Llama Model transformer with a span classification head on top for extractive question-answering tasks like
+SQuAD (a linear layer on top of the hidden-states output to compute `span start logits` and `span end logits`).
+    """,
+    LLAMA_START_DOCSTRING,
+)
+class LlamaForQuestionAnswering(LlamaPreTrainedModel):
+    # Copied from transformers.models.bloom.modeling_bloom.BloomForQuestionAnswering.__init__ with Bloom->Llama
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1431]: what not use this for the entire class?

**Review by ArthurZucker — COMMENTED** [review]
Left a small nit, LGTM otherwise

**Comment by ArthurZucker** [comment]
failing tests are flaky !

**Comment by nakranivaibhav** [comment]
@NielsRogge or @ArthurZucker What is to be done kindly guide further?

**Comment by ArthurZucker** [comment]
Just waiting for you to answer my last comment 😉

**Comment by nakranivaibhav** [comment]
@ArthurZucker Why not use this for the entire class? This one or
failing tests are flaky !
The former i have given some explanation, the latter i have to pray the flaky gods 😁

**Review by ArthurZucker — COMMENTED** [review]
Thanks for you effort. 
Could you: 
1. Update the PR description to something more befitting!
2. Explain why you can't use the copied from statement `# Copied from transformers.models.bloom.modeling_bloom.BloomForQuestionAnswering with Bloom->Llama` at the top of the class? 
3. if the failing CI is unrelated we'll be able to merge (test exotic model or something flaky!)

**Comment by nakranivaibhav** [comment]
@ArthurZucker Alright i will get to work

**Comment by nakranivaibhav** [comment]
@ArthurZucker 
I did update the PR title.

I tried adding `# Copied from transformers.models.bloom.modeling_bloom.BloomForQuestionAnswering with Bloom->Llama` at the top of the class.
Issues I am facing:
It wants the entire class code to be the same as Bloom when I make fix-copies
Bloom has a `head_mask` param in the forward function which Llama does not seem to handle anywhere.
Bloom does not have `past_key_values` in its forward params.
Not including `past_key_values` in llama fails some tests.
I did add `past_key_values` in this commit [c90f9b3] for this reason.

For the tests I have this error 
`FAILED examples/pytorch/test_accelerate_examples.py::ExamplesTestsNoTrainer::test_run_swag_no_trainer - AssertionError: 0.3 not greater than or equal to 0.8`

How do I figure out the flaky tests besides being in the exotic model test?

**Comment by ArthurZucker** [comment]
Alright no worries then no need to use the copied from for everything, thanks for explaining 😉

**Review by ArthurZucker — APPROVED** [review]
LGTM, failing test is unrelated I'll merge

**Comment by ArthurZucker** [comment]
image
Flaky tests are marked as flaky automatically 😉

**Comment by nakranivaibhav** [comment]
@ArthurZucker If the community demands. I can do the same for mistral as well. :)

**Pull request merged**